source/_posts/2015-11-16-immutable-strings-in-ruby-2-dot-3.markdown

```diff
@@ -0,0 +1,279 @@
+  github_handle: axxiss
+  image: /images/team/
+  description: Software Engineer. Turning coffee into software since 2008.
+---
+
+String literals immutability is one of the new features you can find in
+[Ruby 2.3][1]. But, what does this imply?
+
+* Same value literals points to the same object
+* Better performance
+* Thread safety
+
+
+Currently string immutability is included as opt-in, being the first step of a
+multi stage rollout and if it works well it will be
+[enabled by default on Ruby 3.0][2]. So it still can end up being just an
+experiment. Time will tell.
+
+You can follow frozen string literals feature status in Ruby issue tracker
+[Feature #11473: Immutable String literal in Ruby 3][5]
+
+<!--more-->
+
+# Immutability
+
+Quoting [wikipedia][3]:
+
+> An immutable object is an object whose state cannot be modified after it is
+> created.
+
+As immutable objects state cannot be changed it doesn't make sense to copy the
+same object over and over. In fact, Ruby 2.3 will hold a unique instance for
+each string literal value used in the program. This means that 'a' and 'a'
+references the same object.
+
+This approach will help to avoid unnecessary data duplication, therefore
+unnecessary object allocation is prevented and it will imply that the program
+runs faster because it will spend less time in the Garbage Collector.
+
+Furthermore, when we're concurrently accessing a shared literal across several
+threads, there is no need to synchronize it access. This could lead to simpler
+multi-threaded programs.
+
+# Moving towards immutability
+
+On this initial phase of Ruby's strings immutability, we can enable this feature
+at a project level through a flag or on a per file basis using a pragma. Furthermore
+for a more detailed error stack (while debugging) another flag can be enabled.
+
+If you're working on an existent project and want to introduce this new feature,
+obviously the most appropriate method is to use the pragma approach. Just add the
+following comment at the beginning of each file you want to switch to immutable
+strings and your done. Well, not really. Probably you will have some errors to fix.
+
+    # frozen_string_literal: true
+
+But, if you woke up today feeling like a kamikaze, enable frozen strings for the
+whole project and prepare yourself for some serious bug hunting. So, to enable
+string immutability for a project add the following flag when running your app.
+
+    --enable-frozen-string-literal
+
+One way or another, it's quite probable that you will encounter some bugs.
+Mainly because you are relying in 3rd party gems that are not prepared for
+immutable strings. Ruby ecosystem will need some time to adapt to this new
+feature, once most popular gems start to adapt frozen string literals the
+transition will be smoother.
+
+Let's run a quick example, create a file named `replace.rb` with the following
+lines:
+
+{% codeblock lang:ruby %}
+  str = "abcde"
+  str.gsub!("c", "b")
+{% endcodeblock %}
+
```

Comment:
Maybe a tiny introductory phrase saying something like `Even though this code is OK in current Rubies, it's mutating str. So if you opt-in for immutable strings using frozen string literal flag enabled, will throw an error like the one below.`